### PR TITLE
[e2e/ndm] add snmp test environment

### PR DESCRIPTION
**Pull request by pducolin**
### What does this PR do?

Add a test environment setup for Snmp tests, including:

- a VM instance on EC2
- a dockerised agent and snmp sim

snmprec files are copied from [integrations-core](https://github.com/DataDog/integrations-core/tree/master/snmp/tests/compose)

### Motivation

AETT-12

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by pducolin** [comment]
Note for reviewers: commit b285e57 is a copy paste from https://github.com/DataDog/integrations-core/tree/master/snmp/tests/compose

File: test/new-e2e/ndm/snmp/snmpTestEnv.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	if !found {
+		return nil, errors.New("unable to find host ip")
+	}
+	snmpTestEnv.InstanceIP = output.Value.(string)
+
+	return snmpTestEnv, nil
+}
+
+// Close performs cleanup and destroys the infra
+func (e *TestEnv) Close() {
```
Comment (ogaca-dd) [inline test/new-e2e/ndm/snmp/snmpTestEnv.go:125]: question: Is this function needed?

**Review by ogaca-dd — CHANGES_REQUESTED** [review]
Great PR,
I am wondering if the data for the environment creation or for the e2e tests should be located in `datadog-agent` repository. I am afraid many teams to have this need leading to have a lot of data in datadog-agent repository.

File: test/new-e2e/ndm/snmp/snmpTestEnv.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	instanceName := fmt.Sprintf("snmp-agent-%s", name)
+
+	if shouldDestroy {
+		fmt.Println("Starting stack destroy")
+		stackManager.DeleteStack(ctx, envName, instanceName)
+		fmt.Println("Stack successfully destroyed")
+		return snmpTestEnv, nil
+	}
+
+	config := auto.ConfigMap{
```
[the remark this answers is not in the document]
Comment (pducolin) [reply test/new-e2e/ndm/snmp/snmpTestEnv.go:48]: Agreed, should be hidden in further PR

File: test/new-e2e/ndm/snmp/snmpTestEnv.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+				return err
+			}
+			dontUseSudo := false
+			fileCommand, err := vm.FileManager.CopyInlineFile(fileName, pulumi.String(fileContent), path.Join(dataPath, fileName), dontUseSudo,
+				pulumi.DependsOn([]pulumi.Resource{createDataDirCommand}))
+			if err != nil {
+				return err
+			}
+			fileCommands = append(fileCommands, fileCommand)
+		}
```
[the remark this answers is not in the document]
Comment (pducolin) [reply test/new-e2e/ndm/snmp/snmpTestEnv.go:82]: Note that here the folder is embedded, so it not exactly a local folder. `embed` has a `(embed.FS).ReadDir(name string)` that returns a list of `io.fs.DirEntry`, like I use in `loadDataFileNames`

```golang
func loadDataFileNames() (out []string, err error) {
	fileEntries, err := dataFolder.ReadDir(composeDataPath)
	if err != nil {
		return nil, err
	}
	for _, f := range fileEntries {
		out = append(out, f.Name())
	}
	return out, nil
}
```

We could have am API accepting a list of `io.DirEntry`. Or a function in `datadog-agent` that handles the copy of an embedded folder to a remote folder.

File: test/new-e2e/ndm/snmp/snmpTestEnv.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+	return snmpTestEnv, nil
+}
+
+// Close performs cleanup and destroys the infra
+func (e *TestEnv) Close() {
+}
+
+//go:embed compose/data
+var dataFolder embed.FS
```
[the remark this answers is not in the document]
Comment (pducolin) [reply test/new-e2e/ndm/snmp/snmpTestEnv.go:129]: Sure ! This should also reply to your point 

> I am wondering if the data for the environment creation or for the e2e tests should be located in datadog-agent repository. I am afraid many teams to have this need leading to have a lot of data in datadog-agent repository.

These files are required by [snmpsim](https://github.com/etingof/snmpsim) to emulate different devices. An snmprec is a list of `OID | TYPE | VALUE` entries. Each file maps to a different device, and smpsim will load them at compose time. The agent sends `snmp` requests to snmpsim: based on the request parameters, snmpsim will reply with data from one of the snmprec. For example, when snmpsim receives an snmp request with version 2 and community string `foo`, snmpsim will look for `foo.snmprec` file and reply based on its content. In other words, they are a local database of metrics and metadata.

In the current development flow, when we add a feature we often update snmprec while updating end to end tests: here is an example from the latest PR updating tests on integrations-core ---> https://github.com/DataDog/integrations-core/pull/13517/files

We (NDM) considered taking them out of end to end code, this adds some friction when working on code and end to end tests. 

What are the concerns of increasing the `datadog-agent` repo size ? Note these files are dev / test dependencies, they should not be part of the agent build for instance.

File: test/new-e2e/ndm/snmp/snmpTestEnv.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+	upResult, err := infra.GetStackManager().GetStack(ctx, envName, instanceName, config, func(ctx *pulumi.Context) error {
+		// setup VM
+		vm, err := ec2.NewVM(ctx)
+		if err != nil {
+			return err
+		}
+
+		// upload snmpsim data files
+		createDataDirCommand, dataPath, err := vm.FileManager.TempDirectory("data")
```
[the remark this answers is not in the document]
Comment (pducolin) [reply test/new-e2e/ndm/snmp/snmpTestEnv.go:61]: I use `/tmp/data` to use a path where I do not need `sudo`. If I use a fix name folder I would still need to create it first. Or we should have `FileManager` taking care of creating path if it does not exist. What do you think ?

File: test/new-e2e/ndm/snmp/snmpTestEnv.go (hunk 1 above)
Comment (pducolin) [reply test/new-e2e/ndm/snmp/snmpTestEnv.go:125]: Thought it was part of an interface, but I am currently the only one implementing it. Removing it, might add something back later

File: test/new-e2e/ndm/snmp/snmpTestEnv.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+func NewTestEnv(name, keyPairName, ddAPIKey, ddAPPKey string, shouldDestroy bool) (*TestEnv, error) {
+	ctx := context.Background()
+	snmpTestEnv := &TestEnv{}
+
+	stackManager := infra.GetStackManager()
+	envName := "aws/sandbox"
+	instanceName := fmt.Sprintf("snmp-agent-%s", name)
+
+	if shouldDestroy {
```
Comment (FlorianVeaux) [inline test/new-e2e/ndm/snmp/snmpTestEnv.go:41]: **💭 thought:** ‏It feels a weird UX to me that NewTestEnv either creates a stack and run test or delete an existing stack. Should it be two separate commands/entrypoints? For instance `ddev` has different entrypoints to start a new env, test on an existing env and delete an existing env. It also provides a convenient entrypoint for doing all three in a row

File: test/new-e2e/scenarios/ndm/main.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+	"github.com/DataDog/datadog-agent/test/new-e2e/ndm/snmp"
+)
+
+func main() {
+	envNamePtr := flag.String("name", "snmp", "environment name")
+	destroyPtr := flag.Bool("destroy", false, "[optional] should destroy the environment")
+	keyPairNamePtr := flag.String("keyPairName", "", "ssh key pair name. Should be uploaded on the cloud provider")
+	apiKeyPtr := flag.String("apiKey", "", "Datadog API key")
+	appKeyPtr := flag.String("appKey", "", "Datadog APP key")
```
Comment (FlorianVeaux) [inline test/new-e2e/scenarios/ndm/main.go:15]: Not that these should be done as part of this PR, but should end user of the framework have to care about api keys and ssh keys?

File: test/new-e2e/utils/credentials/store-aws.go [hunk 8]
```diff
@@ -21,7 +21,8 @@ func (s *awsStore) get(key string) (string, error) {
 		return "", err
 	}
 
-	output, err := ssmClient.GetParameter(context.Background(), &ssm.GetParameterInput{Name: &key, WithDecryption: true})
+	withDecription := true
+	output, err := ssmClient.GetParameter(context.Background(), &ssm.GetParameterInput{Name: &key, WithDecryption: &withDecription})
```
Comment (FlorianVeaux) [inline test/new-e2e/utils/credentials/store-aws.go:25]: Is this change intended?

**Review by FlorianVeaux — APPROVED** [review]
PR looks great to me and is easy to follow even with little pulumi context, great job!
Added a few questions and thoughts but nothing blocking to me.

File: test/new-e2e/ndm/snmp/snmpTestEnv.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/ndm/snmp/snmpTestEnv.go:82]: Both solutions are fine.

File: test/new-e2e/scenarios/ndm/main.go [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"flag"
+	"fmt"
+
+	"github.com/DataDog/datadog-agent/test/new-e2e/ndm/snmp"
+)
+
+func main() {
+	envNamePtr := flag.String("name", "snmp", "environment name")
+	destroyPtr := flag.Bool("destroy", false, "[optional] should destroy the environment")
```
Comment (ogaca-dd) [inline test/new-e2e/scenarios/ndm/main.go:12]: Note: We need to think how to start an env as having a `main` for each env doesn't scale well.

Comment (pducolin) [reply test/new-e2e/scenarios/ndm/main.go:12]: We should probably have an interface for environments with a "CreateAndStart" and "Destroy" functions called by a common main, with the cool auto-discover you implemented in a previous PR in `test-infra`

File: test/new-e2e/utils/credentials/store-aws.go (hunk 8 above)
Comment (pducolin) [reply test/new-e2e/utils/credentials/store-aws.go:25]: Yes, hard to read, but `WithDecription` moved from type `bool` (boolean value) to type `*bool` (boolean pointer).

File: test/new-e2e/scenarios/ndm/main.go (hunk 7 above)
Comment (pducolin) [reply test/new-e2e/scenarios/ndm/main.go:15]: @vboulineau has a plan to silently upload the default user public key (`id_rsa`) and add it to the remote instance known keys. For DD keys I would like to follow `ddev` and use environment variables that could be overridden with command line args

File: test/new-e2e/ndm/snmp/snmpTestEnv.go (hunk 6 above)
Comment (pducolin) [reply test/new-e2e/ndm/snmp/snmpTestEnv.go:41]: Changed it ! Moreover, I realised you need to first upsert a stack before deleting it - the stack manager needs to load the stack state syncing with remote

File: test/new-e2e/ndm/snmp/snmpTestEnv.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pducolin) [reply test/new-e2e/ndm/snmp/snmpTestEnv.go:80]: Will address this separately, adding a `CopyEmbedFolder` and `CopyFolder` to `test-infra`

[the remark this answers is not in the document]
Comment (ogaca-dd) [reply test/new-e2e/ndm/snmp/snmpTestEnv.go:80]: Track by AETT-47

**Review by ogaca-dd — APPROVED** [review]
👍

**Comment by pducolin** [comment]
Forced push after rebase on main, after https://github.com/DataDog/datadog-agent/pull/13643 merge

File: test/new-e2e/scenarios/ndm/main.go (hunk 7 above)
Comment (vboulineau) [reply test/new-e2e/scenarios/ndm/main.go:20]: It's already done but currently it means you must use `~/.ssh/id_rsa` to connect.

**Review by vboulineau — APPROVED** [review]

**Comment by pducolin** [comment]
Dropped the readme commit

File: test/new-e2e/README.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
+# End to end tests
+
+## Setup
```
[the remark this answers is not in the document]
Comment (pducolin) [reply test/new-e2e/README.md]: Dropped the readme, ok for you if I add details to `test-infra` Readme ?
